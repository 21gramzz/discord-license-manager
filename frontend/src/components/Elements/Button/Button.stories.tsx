import { ComponentMeta, ComponentStoryObj } from '@storybook/react';
import { Button } from './Button';

export default { component: Button } as ComponentMeta<typeof Button>;

export const Default: ComponentStoryObj<typeof Button> = {
  args: {
    children: 'Default Button',
  },
};

export const Primary: ComponentStoryObj<typeof Button> = {
  args: {
    variants: 'primary',
    children: 'Primary Button',
  },
};
