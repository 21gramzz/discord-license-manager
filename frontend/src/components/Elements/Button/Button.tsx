import React from 'react';
import styled from 'styled-components';

export interface ButtonProps {
  className?: string;
  type?: 'button' | 'submit' | 'reset';
  children: React.ReactNode;
  color?: string;
  width?: string;
  height?: string | number;
  margin?: string;
  padding?: string;
  fontSize?: string;
  border?: string;
  backgroundColor?: string;
  variants?: 'default' | 'primary';
  onClick?: () => void;
}

export const Button: React.FC<ButtonProps> = ({
  variants = 'default',
  ...props
}) => {
  return (
    <>
      {variants === 'primary' ? (
        <StyledPrimaryButton {...props}>{props.children}</StyledPrimaryButton>
      ) : (
        <StyledButton {...props}>{props.children}</StyledButton>
      )}
    </>
  );
};

const StyledButton = styled.button<
  Pick<
    ButtonProps,
    | 'margin'
    | 'padding'
    | 'color'
    | 'backgroundColor'
    | 'width'
    | 'fontSize'
    | 'height'
    | 'border'
  >
>`
  user-select: none;
  color: ${({ color, theme }) => color || theme.text};
  background-color: ${({ backgroundColor, theme }) =>
    backgroundColor || theme.surface};
  width: ${({ width }) => width || 'auto'};
  height: ${({ height }) => height || '3.5rem'};
  margin: ${({ margin }) => margin || '0'};
  padding: ${({ padding }) => padding || '0.5rem 1.5rem'};
  box-sizing: border-box;
  outline: 0;
  border: ${({ border, theme }) => border || `solid 0.5px ${theme.border}`};
  font-weight: 400;
  font-size: ${({ fontSize }) => fontSize || '1.3rem'};
  letter-spacing: 0.025em;
  border-radius: 0.5rem;
  text-align: center;
  align-items: center;
  white-space: nowrap;
  display: inline-flex;
  justify-content: center;
  &:active {
    transform: translate(-1px, -1px);
  }
`;

const StyledPrimaryButton = styled(StyledButton)`
  background-color: ${({ theme }) => theme.primary};
  color: white;
  border: 0px;
`;
