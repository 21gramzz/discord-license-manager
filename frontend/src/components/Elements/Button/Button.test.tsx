import initStoryshots from '@storybook/addon-storyshots';
import * as stories from './Button.stories';
import { composeStories } from '@storybook/testing-react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

initStoryshots();

describe('Button', () => {
  const { Default } = composeStories(stories);

  test('should render default button', async () => {
    render(<Default>Click here</Default>);
    expect(screen.getByText(/click here/i)).toBeInTheDocument();
  });

  test('should render primary button', async () => {
    render(<Default variants="primary">Click here</Default>);
    expect(screen.getByText(/click here/i)).toBeInTheDocument();
  });

  test('should call click event', async () => {
    const handleClick = jest.fn();
    render(<Default onClick={handleClick}>Click here</Default>);
    fireEvent.click(screen.getByRole('button'));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
