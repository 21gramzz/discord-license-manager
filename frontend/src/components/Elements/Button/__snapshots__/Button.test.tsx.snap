// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Button Default 1`] = `
<button
  className="sc-bczRLJ bshnOH"
>
  Click here
</button>
`;
